### ISSUE TITLE
Implement new config

### ISSUE THREAD

**Issue by zamotany** [callstack/haul#323]
Historically, the configuration has been done through the file `webpack.haul.js`. This proposal aims to achieve 2 things,

1. Simplify configuring commonly needed features, such as including folders for transpilation
2. Separate webpack configuration so that other tools can use it, e.g. - eslint-resolver-webpack
3. Have flexibility to configure Haul's options independent from Webpack

With the proposed approach, configuration will be done through a file called `haul.config.js`, which will look kinda like this:

```js
export default {
  // haul specific options
  webpack: {
    // webpack configuration
  }
}
```

This allows us to have the flexibility to add additional options which are unrelated to webpack, and also allows us to switch the bundler to a different one in future.

The `webpack` field can be a plain object with webpack configuration, or a function which receives various arguments:

```js
export default {
  // haul specific options
  webpack: ({ platform, production }) => ({
    // return custom config
  });
}
```

This mirrors a vanilla webpack config and it's `env` argument. This makes it so that other tools can understand the webpack config. One could also extract the webpack configuration to a separate file (e.g. - `webpack.config.js`) which other tools can read.

Currently we pass a second argument `defaults`, which is convenient, but makes the config incompatible with tools which require vanilla webpack config. We'll continue to support it, but we can also export a helper utility in case you need to support those tools:

```js
import { createWebpackConfig } from 'haul';

export default {
  webpack: createWebpackConfig(({ platform }) => {
    entry: `./index.${platform}.js`,
    include: /node_modules/,
  }),
}
```

The `createWebpackConfig` utility can take some common options has return a webpack configuration function which takes the `env` argument. To override configurations from a confguration generated this way, you could do something like this:

```js
import { createWebpackConfig } from 'haul';

export default {
  webpack: env => {
    const config = createWebpackConfig({
      entry: `./index.js`,
      include: /node_modules/,
    })(env);

    config.plugins.push(new CaseSensitivePathsPlugin());

    return config;
  },
}
```

**Comment by hedgepigdaniel** [comment]
This looks useful - e.g. one useful option outside the webpack config might be to customize the webpack stats output.

**Comment by simonbuchan** [comment]
So to be clear, one could do:

```js
import webpackConfig from "./webpack.config";
export default {
  webpack: env => webpackConfig({ haul: env }),
};
```

to have a single webpack config file? And this would not munge the final output, such as injecting polyfills (e.g. if I'm already injecting `core-js`), babel 6 with `react-native` preset etc?

**Comment by zamotany** [comment]
@simonbuchan 

Not really, [our Webpack config](https://github.com/callstack/haul/blob/master/src/utils/makeReactNativeConfig.js#L61-L206) is specifically tailored for React Native, with custom loader, resolvers and other tweaks, so unless you want to recreate it in your config and maintain it yourself, you would need to use `createWebpackConfig` from `haul`.

As far as polyfills are concerned, we don't use `core-js` nor `babel-polyfill`. Instead we have our [custom one](https://github.com/callstack/haul/blob/master/src/utils/polyfillEnvironment.js), created specifically for RN.

**Comment by simonbuchan** [comment]
I am aware. Having just spent a bunch of time trying to get haul to work with my existing code and webpack config 😿

Consider optimizing for being able to share with "standard" webpack configs, since that's a very useful use-case. In particular, having a "node_modules/react-native only" config that won't try to transpile user code would be nice, since it seems like you could simply use `webpack-merge` to combine it with your own code.

**Comment by jukben** [comment]
As a team member of our Callstack Open Source division – I'm on it. 💪

**Comment by thymikee** [comment]
Fixed via #372
